tutorials/CountAndCache.qmd

```diff
@@ -122,3 +123,222 @@ result as usual:
+```
+
+But since both the cost and the gradient require the computation of the matrix-vector product $Ap$, it might be beneficial to only compute this once.
+
+### The [`ManifoldCostGradientObjective`](@ref) approach
+
+The [`ManifoldCostGradientObjective`](@ref) uses a combined function to compute both the gradient and the cost at the same time. We define the inplace variant as
+
+```{julia}
+function g_grad_g!(M, X, p)
+    X .= -A*p
+    c = p'*X
+    X .*= 2
+    project!(M, X, p, X)
+    return (c, X)
+end
+```
+
+where we only compute the matrix-vector product once.
+The small disadvantage might be, that we always compute _both_, the gradient and the cost. Luckily, the cache we used before, takes this into account and caches both results, such that we indeed end up computing `A*p` only once when asking to a cost and a gradient.
+
+Let's compare both methods
+```{julia}
+p0 = [(1/5 .* ones(5))..., zeros(m-4)...];
+@time s1 = gradient_descent(N, g, grad_g!, p0;
+    stopping_criterion = StopWhenGradientNormLess(1e-5),
+    evaluation=InplaceEvaluation(),
+    count=[:Cost, :Gradient],
+    cache=(:LRU, [:Cost, :Gradient], 25),
+    return_objective=true,
+)
+```
+
+versus
+
+```{julia}
+obj = ManifoldCostGradientObjective(g_grad_g!; evaluation=InplaceEvaluation())
+@time s2 = gradient_descent(N, obj, p0;
+    stopping_criterion=StopWhenGradientNormLess(1e-5),
+    count=[:Cost, :Gradient],
+    cache=(:LRU, [:Cost, :Gradient], 25),
+    return_objective=true,
+)
+```
+
+first of all both yield the same result
+
+```{julia}
+p1 = get_solver_result(s1)
+p2 = get_solver_result(s2)
+[distance(N, p1, p2), g(N, p1), g(N, p2), f_star]
+```
+
+and we can see that the combined number of evaluations is once 2051, once
+just the number of cost evaluations 1449. Note that the involved additional 847
+gradient evaluations are merely a multiplication with 2. On the other hand, the additional caching of the gradient in these cases might be less beneficial.
+It is beneficial, when the gradient and the cost are very often required together.
+
+### A shared storage approach using a functor
+
+So compared to the previous approach, which might be beneficial.
+An alternative is the usage of a functor, that introduces a “shared storage”
+of the result of computing `A*p`. We additionally have to store `p` though,
+since we have to check that we are still evaluation the cost and/or gradient
+at the same point, respectively.
+We again consider the (more efficient) inplace variant.
+The can be done as follows
+
+```{julia}
+struct StorageG{T,M}
+    A::M
+    Ap::T
+    p::T
+end
+function (g::StorageG)(::Val{:Cost}, M, p)
+    if !(p==g.p) #We are at a new point -> Update
+        g.Ap .= g.A*p
+        g.p .= p
+    end
+    return -g.p'*g.Ap
+end
+function (g::StorageG)(::Val{:Gradient}, M, X, p)
+    if !(p==g.p) #We are at a new point -> Update
+        g.Ap .= g.A*p
+        g.p .= p
+    end
+    copyto!(M, X, p, -2.0 .* g.Ap)
```

Comment:
```suggestion
    X .= -2 .* g.Ap
```

Reply:
I am never sure when to use which, so I trust you here.